File: packages/apigateway/config/sport_config.go
// Package config provides sport configuration management for the API Gateway.
//
// # Design: Singleton Loading + Dependency Injection for Usage
//
// This package uses a two-layer approach:
//
//  1. Loading layer: LoadSportConfig() uses sync.Once to load config exactly once.
//     This is appropriate because sport config is application-wide, immutable after
//     startup, and expensive to parse/validate.
//
//  2. Usage layer: Handlers and business logic receive *SportConfig via constructor
//     injection (see activities.NewHandler). This enables proper testing and avoids
//     hidden dependencies in business logic.
//
// For testing:
//   - Unit tests that need custom configs use loadSportConfigInternal() directly
//   - Handler tests create configs and inject them via constructors
//   - The sync.Once doesn't affect test isolation because handlers use injected instances
package config

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"slices"
	"sync"

	"github.com/go-playground/validator/v10"
)

// UnknownSportCategory is the fallback category returned when a Strava sport_type
// has no explicit mapping in sport_types.json. The catch-all keeps unmapped
// activities visible in the UI ("Other") instead of silently dropping them.
//
// Operators are alerted via the "Unknown Strava sport_type detected" log-based
// metric (see terraform/modules/desirelines/alerts.tf). When the alert fires,
// add the upstream sport to schemas/sports/sport_types.json so it lands in its
// proper category.
const UnknownSportCategory = "other"

// unknownSportLogMessage is the human-readable log message for unmapped
// sport_type values. It is NO LONGER what the metric keys off (see
// unknownSportLogEvent), so it can be reworded freely.
const unknownSportLogMessage = "Unknown Strava sport_type detected"

// unknownSportLogEvent is the stable structured "event" field the GCP
// log-based metric (terraform/modules/desirelines/monitoring.tf) filters on:
// jsonPayload.event="unknown_sport_type". Keying the metric on this machine
// field instead of the prose message means a reworded message can't silently
// break the alert. It is a MONITORING CONTRACT shared across two runtimes —
// keep byte-identical with the Python emitter (UNKNOWN_SPORT_LOG_EVENT) and the
// terraform filter; pinned by TestGetCategoryForStravaType_UnknownLogsWarning.
const unknownSportLogEvent = "unknown_sport_type"

//go:embed sport_types.json
var embeddedSportConfig []byte

// SportCategory defines a sport category and its mapping to Strava's data model.
//
// # Strava type vs sport_type
//
// Strava's API has two activity classification fields:
//   - type: broad/deprecated category (e.g., "Workout" covers yoga, weight training, HIIT, etc.)
//   - sport_type: specific activity kind (e.g., "Yoga", "WeightTraining", "HIIT")
//
// StravaTypes contains sport_type values (the specific ones), NOT type values.
// The database stores both: column 'type' = Strava type, column 'sport' = Strava sport_type.
// All filtering queries must use the 'sport' column to match against StravaTypes.
type SportCategory struct {
	DisplayName string `json:"displayName" validate:"required"`
	// StravaTypes lists Strava sport_type values belonging to this category.
	// Example: cycling = ["Ride", "VirtualRide", "GravelRide", "MountainBikeRide", ...]
	// These match the DB 'sport' column, NOT the 'type' column.
	StravaTypes   []string `json:"stravaTypes" validate:"required,min=1"`
	ExcludedTypes []string `json:"excludedTypes"`
	PrimaryMetric string   `json:"primaryMetric" validate:"required"`
	Metrics       []string `json:"metrics" validate:"required,min=1"`
	HasDistance   bool     `json:"hasDistance"`
	HasElevation  bool     `json:"hasElevation"`
	// DangerPace is the optional sustainable-pace limit used by the frontend's
	// "danger zone" rendering. Loaded and passed through verbatim; the apigateway
	// does not interpret it.
	DangerPace *DangerPace `json:"dangerPace,omitempty"`
	// GoalDefaults is optional per-sport goal tuning (increment / rounding /
	// default / chart intervals) used by the frontend's metric-config layer.
	// Loaded and passed through verbatim; the apigateway does not interpret it.
	GoalDefaults *GoalDefaults `json:"goalDefaults,omitempty"`
}

// DangerPace expresses a daily pace ceiling in human-readable units. The web
// client converts it to the user's preferred display unit before comparing.
type DangerPace struct {
	ValuePerDay float64 `json:"valuePerDay" validate:"required"`
	Unit        string  `json:"unit" validate:"required"`
	// WarnAtFraction is the optional fraction of the threshold at which the
	// frontend's pacing chart starts showing the "danger zone" overlay
	// (defaults to 0.75 when omitted). Loaded and passed through verbatim, but
	// range-validated at startup so a misconfigured registry fails fast.
	WarnAtFraction *float64 `json:"warnAtFraction,omitempty" validate:"omitempty,gte=0,lte=1"`
}

// GoalDefaults is optional per-sport goal tuning consumed by the frontend. Each
// field is optional and inherits the base metric config when omitted. Loaded
// and passed through verbatim; the apigateway does not interpret it.
type GoalDefaults struct {
	Increment      *float64        `json:"increment,omitempty"`
	Rounding       *float64        `json:"rounding,omitempty"`
	DefaultValue   *float64        `json:"defaultValue,omitempty"`
	ChartIntervals []ChartInterval `json:"chartIntervals,omitempty"`
}

// ChartInterval is one Y-axis tick threshold. A nil Max marks the catch-all
// top bucket (JSON has no Infinity; the frontend restores it).
type ChartInterval struct {
	Max      *float64 `json:"max,omitempty"`
	Interval float64  `json:"interval"`
}

// SportConfigData is a config that holds the sport config version and
//
//	the valid sport categories the application can serve
type SportConfigData struct {
	Version         string                   `json:"version" validate:"required"`
	SportCategories map[string]SportCategory `json:"sportCategories" validate:"required,min=1,dive"`
}

// SportConfig contains the data as a SportConfigData struct
type SportConfig struct {
	data SportConfigData
	// reverseMap maps Strava sport_type values to category names.
	// Built at load time for O(1) lookups. E.g., "Ride" → "cycling".
	reverseMap map[string]string
	// rawJSON stores the original JSON bytes used to create this config.
	// Used for serving the config via API endpoint.
	rawJSON []byte
	// unknownSeen deduplicates the WARNING log emitted by
	// GetCategoryForStravaType so a single unmapped sport_type doesn't spam
	// logs across a request burst. Cleared with the process — Cloud Run
	// recycling naturally re-arms the dedup, which is exactly what the
	// log-based alert wants (one fresh sighting per restart).
	unknownSeen sync.Map
}

// Package-level state for singleton loading pattern.
// These are only used by LoadSportConfig() - business logic receives *SportConfig
// via dependency injection and doesn't access these directly.
var (
	sportConfig     *SportConfig
	sportConfigErr  error
	sportConfigOnce sync.Once
	validate        = validator.New()
)

// SupportedConfigVersions is a list of supported versions
var SupportedConfigVersions = []string{"1.0"}

// LoadSportConfig loads available application SportConfig
// Uses sync.Once to ensure it's only loaded once in production
func LoadSportConfig(configPath string) (*SportConfig, error) {
	sportConfigOnce.Do(func() {
		sportConfig, sportConfigErr = NewSportConfig(configPath)
	})
	return sportConfig, sportConfigErr
}

// NewSportConfig creates a new SportConfig instance from a file or embedded default.
// Exposed for testing to allow creating isolated config instances.
func NewSportConfig(configPath string) (*SportConfig, error) {
	var data []byte
	var err error

	// If path provided, use it (for testing with custom configs)
	// Otherwise, use embedded config
	if configPath != "" {
		// Use file system path (for development/testing with custom configs)
		// #nosec G304 - configPath is a known configuration file path
		data, err = os.ReadFile(configPath)
		if err != nil {
			return nil, fmt.Errorf("failed to read sport config: %w", err)
		}
	} else {
		// Use embedded config (production default)
		if len(embeddedSportConfig) == 0 {
			return nil, fmt.Errorf("embedded sport config not available and no path provided")
		}
		data = embeddedSportConfig
	}

	var configData SportConfigData
	err = json.Unmarshal(data, &configData)
	if err != nil {
		return nil, fmt.Errorf("failed to parse sport config: %w", err)
	}

	// Validate schema with struct tags
	err = validate.Struct(configData)
	if err != nil {
		return nil, fmt.Errorf("invalid sport config schema: %w", err)
	}

	// Validate version (fail fast)
	if !slices.Contains(SupportedConfigVersions, configData.Version) {
		return nil, fmt.Errorf(
			"unsupported sport config version: %s (supports: %v). "+
				"Update application code or rollback config version",
			configData.Version, SupportedConfigVersions,
		)
	}

	// Build reverse lookup map: Strava sport_type → category name.
	// Fail fast if a sport_type appears under more than one category: the map
	// range order is randomized per process, so a collision would otherwise
	// resolve to whichever category was visited last — silently miscategorizing
	// and changing the answer on every restart.
	reverseMap := make(map[string]string)
	for categoryName := range configData.SportCategories {
		// Index the map for just the slice we need rather than ranging the
		// value: `k, v := range` would copy the whole SportCategory struct each
		// iteration (gocritic rangeValCopy) now that it carries optional config.
		stravaTypes := configData.SportCategories[categoryName].StravaTypes
		for _, stravaType := range stravaTypes {
			if existing, dup := reverseMap[stravaType]; dup {
				return nil, fmt.Errorf(
					"invalid sport config: sport_type %q maps to multiple categories "+
						"(%q and %q); each sport_type must belong to exactly one category",
					stravaType, existing, categoryName,
				)
			}
			reverseMap[stravaType] = categoryName
		}
	}

	return &SportConfig{data: configData, reverseMap: reverseMap, rawJSON: data}, nil
}

// ListSports returns all available sports
func (c *SportConfig) ListSports() []string {
	sports := make([]string, 0, len(c.data.SportCategories))
	for sport := range c.data.SportCategories {
		sports = append(sports, sport)
	}
	return sports
}

// GetCategory returns the SportCategory config for a category name.
// For example, GetCategory("cycling") returns the cycling config with its Strava types, metrics, etc.
// Returns false if the category doesn't exist.
func (c *SportConfig) GetCategory(category string) (SportCategory, bool) {
	cat, ok := c.data.SportCategories[category]
	return cat, ok
}

// ValidateSport ensures that an input sport is a sport that
//
//	can be served
func (c *SportConfig) ValidateSport(sport string) bool {
	_, ok := c.data.SportCategories[sport]
	return ok
}

// GetStravaTypes returns the Strava sport_type values that map to a category.
// For example, "cycling" returns ["Ride", "VirtualRide"].
// Returns nil if the category doesn't exist.
func (c *SportConfig) GetStravaTypes(category string) []string {
	cat, ok := c.data.SportCategories[category]
	if !ok {
		return nil
	}
	return cat.StravaTypes
}

// GetCategoryForStravaType returns the category name for a Strava sport_type value.
// For example, "Ride" returns "cycling", "TrailRun" returns "running".
//
// When the sport_type has no mapping (e.g., Strava added a new SportType enum
// value upstream that we haven't registered yet), it returns
// [UnknownSportCategory] ("other") and emits a structured WARNING log so the
// log-based metric "${var.project_name}_${var.environment}_unknown_sport_type"
// can fire an alert. The warning is deduplicated per-process (sync.Map) to
// avoid log spam from request bursts referencing the same unmapped type.
func (c *SportConfig) GetCategoryForStravaType(stravaType string) string {
	if category, ok := c.reverseMap[stravaType]; ok {
		return category
	}
	if stravaType == "" {
		// Empty input is not a real Strava sport_type — bucket it as "other"
		// without logging (would be noisy if a column ever lands NULL).
		return UnknownSportCategory
	}
	if _, loaded := c.unknownSeen.LoadOrStore(stravaType, struct{}{}); !loaded {
		slog.Default().Warn(unknownSportLogMessage,
			"event", unknownSportLogEvent,
			"unmapped_sport_type", stravaType,
			"fallback_category", UnknownSportCategory,
		)
	}
	return UnknownSportCategory
}

// RawJSON returns the raw sport config JSON bytes.
// Used for serving the config via API endpoint.
func (c *SportConfig) RawJSON() []byte {
	return c.rawJSON
}

File: packages/stravapipe/src/stravapipe/config/sport_config.py
"""Sport configuration loader with version and schema validation."""

from functools import lru_cache
import json
import logging
from pathlib import Path
from typing import Any

from pydantic import BaseModel, Field, ValidationError

logger = logging.getLogger(__name__)

# Update when code supports new versions
SUPPORTED_CONFIG_VERSIONS = ["1.0"]

# UNKNOWN_SPORT_CATEGORY is the fallback bucket returned by
# SportConfig.categorize_activity for any Strava sport_type that has no
# explicit mapping in sport_types.json. Mirrors apigateway's
# config.UnknownSportCategory ("other"); changes must stay in lockstep.
UNKNOWN_SPORT_CATEGORY = "other"

# UNKNOWN_SPORT_LOG_MESSAGE is the canonical WARNING message the GCP
# log-based metric filter is bound to (see
# terraform/modules/desirelines/monitoring.tf). Renaming this string
# without also updating Terraform silently breaks the alert.
UNKNOWN_SPORT_LOG_MESSAGE = "Unknown Strava sport_type detected"


class DangerPaceModel(BaseModel):
    """Optional sustainable-pace limit; consumed by the web client only."""

    value_per_day: float = Field(alias="valuePerDay")
    unit: str

    model_config = {"populate_by_name": True}


# Pydantic models for schema validation
class SportCategoryModel(BaseModel):
    """Schema for sport category configuration."""

    display_name: str = Field(alias="displayName")
    # Strava sport_type values (NOT the broad 'type' field).
    # E.g., ["Yoga"] not ["Workout"]. See SportCategory docstring for details.
    strava_types: list[str] = Field(alias="stravaTypes", min_length=1)
    excluded_types: list[str] = Field(default=[], alias="excludedTypes")
    primary_metric: str = Field(alias="primaryMetric")
    metrics: list[str] = Field(min_length=1)
    has_distance: bool = Field(alias="hasDistance")
    has_elevation: bool = Field(alias="hasElevation")
    # Loaded and ignored by stravapipe; web frontend consumes it for danger-zone rendering.
    danger_pace: DangerPaceModel | None = Field(default=None, alias="dangerPace")

    model_config = {"populate_by_name": True}


class SportConfigModel(BaseModel):
    """Schema for sport configuration file."""

    version: str
    sport_categories: dict[str, SportCategoryModel] = Field(
        alias="sportCategories", min_length=1
    )

    model_config = {"populate_by_name": True}


class SportCategory:
    """Sport category configuration.

    Strava has two activity classification fields:
      - type: broad/deprecated category (e.g., "Workout" covers yoga, weights, HIIT)
      - sport_type: specific activity kind (e.g., "Yoga", "WeightTraining", "HIIT")

    strava_types contains sport_type values (the specific ones), NOT type values.
    The DB stores both: column 'type' = Strava type, column 'sport' = Strava sport_type.
    """

    def __init__(self, name: str, config: dict[str, Any]):
        self.name = name
        self.display_name = config["display_name"]
        self.strava_types = set(config["strava_types"])
        self.excluded_types = set(config.get("excluded_types", []))
        self.primary_metric = config["primary_metric"]
        self.metrics = config["metrics"]
        self.has_distance = config["has_distance"]
        self.has_elevation = config["has_elevation"]

    def matches(self, sport_type: str) -> bool:
        """Check if a Strava sport_type value belongs to this category."""
        return sport_type in self.strava_types and sport_type not in self.excluded_types


class SportConfig:
    """Sport configuration manager."""

    def __init__(self, config_path: Path, *, validate_version: bool = True):
        """Load and validate sport configuration.

        Args:
            config_path: Path to sport_types.json file.
            validate_version: If True, raise ValueError for unsupported versions.
                Set to False for testing version handling without triggering errors.
        """
        with config_path.open() as f:
            data = json.load(f)

        # Validate schema with Pydantic
        try:
            validated = SportConfigModel.model_validate(data)
        except ValidationError as e:
            raise ValueError(f"Invalid sport config schema:\n{e}") from e

        self.version = validated.version

        # Validate version (fail fast, consistent with Go implementation)
        if validate_version and self.version not in SUPPORTED_CONFIG_VERSIONS:
            raise ValueError(
                f"Unsupported sport config version: {self.version}\n"
                f"This code supports: {SUPPORTED_CONFIG_VERSIONS}\n"
                f"Update application code or rollback config version."
            )

        self.categories = {
            name: SportCategory(name, config.model_dump())
            for name, config in validated.sport_categories.items()
        }

        # Fail fast if a sport_type appears under more than one category. Python
        # would otherwise resolve a collision to the first match by document
        # order (silent), while the Go loader picks a per-restart-random winner;
        # either way it's a config bug, so reject it at load time in both.
        seen_strava_types: dict[str, str] = {}
        for name, category in self.categories.items():
            for strava_type in category.strava_types:
                existing = seen_strava_types.get(strava_type)
                if existing is not None:
                    raise ValueError(
                        f"Invalid sport config: sport_type {strava_type!r} maps to "
                        f"multiple categories ({existing!r} and {name!r}); each "
                        "sport_type must belong to exactly one category."
                    )
                seen_strava_types[strava_type] = name
        # Per-process dedup for the "Unknown Strava sport_type detected"
        # WARNING. Cloud Run recycles restore the alert's signal naturally —
        # one fresh sighting per restart is the desired behaviour.
        self._unknown_seen: set[str] = set()

    def categorize_activity(self, sport_type: str) -> str:
        """Map a Strava sport_type value to its sport category name.

        Returns the explicit category for known sport_types, or
        ``UNKNOWN_SPORT_CATEGORY`` ("other") as a fallback for any value not
        in ``sport_types.json``. Unknown values trigger a structured WARNING
        log on first sighting (deduplicated per process) that the GCP
        log-based metric pivots on for alerting.

        Args:
            sport_type: Strava ``sport_type`` value (e.g., ``"Ride"``, ``"Run"``).

        Returns:
            Category name (e.g., ``"cycling"``) or ``"other"`` for unmapped
            input. Empty/None-like input falls into ``"other"`` silently
            (no WARNING) to avoid alert noise from NULL columns.
        """
        for name, category in self.categories.items():
            if category.matches(sport_type):
                return name
        if not sport_type:
            return UNKNOWN_SPORT_CATEGORY
        if sport_type not in self._unknown_seen:
            self._unknown_seen.add(sport_type)
            logger.warning(
                UNKNOWN_SPORT_LOG_MESSAGE,
                extra={
                    "unmapped_sport_type": sport_type,
                    "fallback_category": UNKNOWN_SPORT_CATEGORY,
                },
            )
        return UNKNOWN_SPORT_CATEGORY

    def get_category(self, sport: str) -> SportCategory | None:
        """Get configuration for a sport category."""
        return self.categories.get(sport)

    def list_sports(self) -> list[str]:
        """Get list of all configured sport categories."""
        return list(self.categories.keys())


@lru_cache(maxsize=1)
def load_sport_config() -> SportConfig:
    """Load sport configuration (cached, validated).

    Returns:
        SportConfig instance with validated schema and version.

    Raises:
        FileNotFoundError: If sport_types.json is missing.
        ValueError: If schema is invalid or version is unsupported.
    """
    config_path = Path(__file__).parent / "sport_types.json"

    if not config_path.exists():
        raise FileNotFoundError(
            f"Sport config not found: {config_path}\nRun: make sync-sport-config"
        )

    # Version validation happens in SportConfig.__init__
    return SportConfig(config_path)
